geoengine/layers.py

```diff
@@ -0,0 +1,660 @@
+
+        buf.write("<table>")
+        buf.write('<thead><tr><th colspan="2">Layer</th></tr></thead>')
+        buf.write("<tbody>")
+        buf.write(f"<tr><th>name</th><td>{self.name}</td></tr>")
+        buf.write(f"<tr><th>description</th><td>{self.description}</td></tr>")
+        buf.write(f"<tr><th>id</th><td>{self.layer_id}</td></tr>")
+        buf.write(f"<tr><th>provider id</th><td>{self.provider_id}</td></tr>")
+
+        # TODO: better representation of workflow, symbology, properties, metadata
+        buf.write('<tr><th>workflow</th><td align="left">')
+        buf.write(f'<pre>{json.dumps(self.workflow, indent=4)}{os.linesep}</pre></td></tr>')
+        buf.write('<tr><th>symbology</th>')
+        buf.write(f'<td align="left">{json.dumps(self.symbology, indent=4)}{os.linesep}</td></tr>')
+        buf.write(f"<tr><th>properties</th><td>{self.properties}{os.linesep}</td></tr>")
+        buf.write(f"<tr><th>metadata</th><td>{self.metadata}{os.linesep}</td></tr>")
+
+        buf.write("</tbody>")
+        buf.write("</table>")
+
+        return buf.getvalue()
+
+
+# TODO: test
+def layer_collection(layer_collection_id: Optional[LayerCollectionId] = None,
+                     layer_provider_id: LayerProviderId = LAYER_DB_PROVIDER_ID,
+                     offset: int = 0,
+                     limit: int = 20,
+                     timeout: int = 60) -> LayerCollection:
+    '''
+    Retrieve a layer collection that contains layers and layer collections.
+    '''
+
+    session = get_session()
+
+    request = '/layers/collections' if layer_collection_id is None \
+        else f'/layers/collections/{layer_provider_id}/{layer_collection_id}'
+
+    response = req.get(
+        f'{session.server_url}{request}?offset={offset}&limit={limit}',
+        headers=session.admin_or_normal_auth_header,
+        timeout=timeout,
+    )
+
+    if not response.ok:
+        raise GeoEngineException(response.json())
+
+    return LayerCollection.from_response(response.json())
+
+
+# TODO: test
+def layer(layer_id: LayerId,
+          layer_provider_id: LayerProviderId = LAYER_DB_PROVIDER_ID,
+          timeout: int = 60) -> Layer:
+    '''
+    Retrieve a layer from the server.
+    '''
+
+    session = get_session()
+
+    response = req.get(
+        f'{session.server_url}/layers/{layer_provider_id}/{layer_id}',
+        headers=session.admin_or_normal_auth_header,
+        timeout=timeout,
+    )
+
+    if not response.ok:
+        raise GeoEngineException(response.json())
+
+    return Layer.from_response(response.json())
+
+
+def _delete_layer_from_collection(collection_id: LayerCollectionId,
+                                  layer_id: LayerId,
+                                  timeout: int = 60) -> None:
+    '''Delete a layer from a collection'''
+
+    session = get_session()
+
+    response = req.delete(
+        f'{session.server_url}/layerDb/collections/{collection_id}/layers/{layer_id}',
+        headers=session.admin_auth_header,
```

Comment:
maybe throw an error if it is not set instead of doing the request? also at the other places.

Reply:
The property `admin_auth_header` throws an error if it is unset.